flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,387 @@
+
+		FlinkFuture<Object> flinkFuture = new FlinkFuture<>(response);
+
+		return flinkFuture.thenApply(new ApplyFunction<Object, JobStatus>() {
+			@Override
+			public JobStatus apply(Object value) {
+				if (value instanceof JobManagerMessages.CurrentJobStatus) {
+					return ((JobManagerMessages.CurrentJobStatus) value).status();
+				} else if (value instanceof JobManagerMessages.JobNotFound) {
+					throw new RuntimeException(
+						new IllegalStateException("Could not find job with JobId " + jobId));
+				} else {
+					throw new RuntimeException(
+						new IllegalStateException("Unknown JobManager response of type " + value.getClass()));
+				}
+			}
+		});
+	}
+
+	private static class UnboundedSource implements SourceFunction<String> {
+		private boolean running = true;
+
+		@Override
+		public void run(SourceContext<String> ctx) throws Exception {
+			while (running) {
+				ctx.collect("hello");
+				// don't overdo it ... ;-)
+				Thread.sleep(50);
+				if (CheckpointBlockingFunction.afterMessWithZooKeeper.get()) {
+					break;
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			running = false;
+		}
+	}
+
+	private static class CheckpointBlockingFunction
+			extends RichMapFunction<String, String>
+			implements CheckpointedFunction {
+
+		// verify that we only call initializeState()
+		// once with isRestored() == false. All other invocations must have isRestored() == true. This
+		// verifies that we don't restart a job from scratch in case the CompletedCheckpoints can't
+		// be read.
+		static AtomicInteger allowedInitializeCallsWithoutRestore = new AtomicInteger(1);
+
+		// we count when we see restores that are not allowed. We only
+		// allow restores once we messed with the HA directory and moved it back again
+		static AtomicInteger illegalRestores = new AtomicInteger(0);
+		static AtomicInteger successfulRestores = new AtomicInteger(0);
+
+		// whether we are after the phase where we messed with the ZooKeeper HA directory, i.e.
+		// whether it's now ok for a restore to happen
+		static AtomicBoolean afterMessWithZooKeeper = new AtomicBoolean(false);
+
+		static AtomicBoolean failedAlready = new AtomicBoolean(false);
+
+		// also have some state to write to the checkpoint
+		private final ValueStateDescriptor<String> stateDescriptor =
+			new ValueStateDescriptor<>("state", StringSerializer.INSTANCE);
+
+		@Override
+		public String map(String value) throws Exception {
+			getRuntimeContext().getState(stateDescriptor).update("42");
+			return value;
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext context) throws Exception {
+			if (context.getCheckpointId() > 5) {
+				waitForCheckpointLatch.trigger();
+				failInCheckpointLatch.await();
+				if (!failedAlready.getAndSet(true)) {
+					throw new RuntimeException("Failing on purpose.");
+				}
+			}
+		}
+
+		@Override
+		public void initializeState(FunctionInitializationContext context) {
+			if (!context.isRestored()) {
```

Comment:
I am wondering if you can get rid of the `allowedInitializeCallsWithoutRestore` completely and simply make this an `assertEquals(afterMessWithZooKeeper, context.isRestored())`.